src/oracle/OracleWriter.sol

```diff
@@ -1,11 +1,69 @@
+import { stdMath } from "forge-std/Test.sol";
+
+import { IOracleWriter } from "src/interfaces/IOracleWriter.sol";
+import { FactoryStoreLib } from "src/libraries/FactoryStore.sol";
+import { Bytes32Lib } from "src/libraries/Bytes32.sol";
+import { LogCompression } from "src/libraries/LogCompression.sol";
+import { Pair } from "src/Pair.sol";
+import { GenericFactory } from "src/GenericFactory.sol";
+
+abstract contract OracleWriter is Pair, IOracleWriter {
+    using FactoryStoreLib for GenericFactory;
+    using Bytes32Lib for bytes32;
+
+    // 100 basis points per second which is 60% per minute
+    uint256 internal constant MAX_CHANGE_PER_SEC = 0.01e18;
 
-abstract contract OracleWriter is IOracleWriter {
     Observation[65536] public observations;
     uint16 public index = type(uint16).max;
 
+    // maximum allowed rate of change of price per second
+    // to mitigate oracle manipulation attacks in the face of post-merge ETH
+    uint256 public allowedChangePerSecond;
+    uint256 public prevClampedPrice;
+
+    constructor() {
+        uint256 lAllowedChangePerSecond = factory.read("Shared::allowedChangePerSecond").toUint256();
+        require(0 < lAllowedChangePerSecond && lAllowedChangePerSecond <= MAX_CHANGE_PER_SEC, "OW: INVALID_CHANGE_PER_SECOND");
+        allowedChangePerSecond = lAllowedChangePerSecond;
+    }
+
+    function setAllowedChangePerSecond(uint256 aAllowedChangePerSecond) external onlyFactory {
+        require(0 < aAllowedChangePerSecond && aAllowedChangePerSecond <= MAX_CHANGE_PER_SEC, "OW: INVALID_CHANGE_PER_SECOND");
+        emit AllowedChangePerSecondChanged(allowedChangePerSecond, aAllowedChangePerSecond);
+        allowedChangePerSecond = aAllowedChangePerSecond;
+    }
+
+    function _calcClampedPrice(
+        uint256 aCurrRawPrice, uint256 aPrevClampedPrice, uint256 aTimeElapsed
+    ) internal virtual returns (uint256 rClampedPrice, int112 rClampedLogPrice) {
+        if (aPrevClampedPrice == 0) {
+            return (aCurrRawPrice, int112(LogCompression.toLowResLog(aCurrRawPrice)));
+        }
+
+        if (_calcPercentageDiff(aCurrRawPrice, aPrevClampedPrice) > allowedChangePerSecond * aTimeElapsed) {
+            // clamp the price
+            if (aCurrRawPrice > aPrevClampedPrice) {
+                rClampedPrice = aPrevClampedPrice * (1e18 + (allowedChangePerSecond * aTimeElapsed)) / 1e18;
+            }
+            else {
+                assert(aPrevClampedPrice > aCurrRawPrice);
+                rClampedPrice = aPrevClampedPrice * (1e18 - (allowedChangePerSecond * aTimeElapsed)) / 1e18;
+            }
+            rClampedLogPrice = int112(LogCompression.toLowResLog(rClampedPrice));
+        }
+        else {
+            rClampedPrice = aCurrRawPrice;
+            rClampedLogPrice = int112(LogCompression.toLowResLog(aCurrRawPrice));
+        }
+    }
+
+    function _calcPercentageDiff(uint256 a, uint256 b) private pure returns (uint256) {
+        return stdMath.percentDelta(a, b);
+    }
```

Comment:
Can you make a test case that tests the threshold +- 1 so that we can detect if `percentDelta` ever changes behaviour?

Reply:
okay created a few basic tests. Only with the `uint` function since that's what we are using.

Also discovered that the order of arguments matters. The base of comparison should be `b` which is what we are doing

Reply:
Those are good but I also meant tests of the oracle writer tolerance. So that regardless of what internal implementation we use we have locked in the observable behavior to be how we expect/want it to be.

Reply:
okay I wrote two cases at the edges. Please see if they're sound